File: src/ClientAPI/Private/MikanResponseFuture.cpp
#include "MikanResponseFuture.h"
#include "MikanRequestManager.h"

struct MikanResponseFutureImpl
{
	MikanRequestManager* ownerRequestManager = nullptr;
	MikanRequestID requestId= INVALID_MIKAN_ID;
	std::future<MikanResponsePtr> future;
};

MikanResponseFuture::MikanResponseFuture() : m_impl(nullptr)
{
}

MikanResponseFuture::MikanResponseFuture(MikanAPIResult result)
	: m_impl(new MikanResponseFutureImpl())
{
	// Immediately set the result on the future
	MikanResponsePromise promise;
	m_impl->future = promise.get_future();
	promise.set_value(makeSimpleMikanResponse(result));
}

MikanResponseFuture::MikanResponseFuture(
	MikanRequestManager* owner, 
	MikanRequestID requestId, 
	MikanResponsePromise& promise)
	: m_impl(new MikanResponseFutureImpl())
{
	m_impl->ownerRequestManager= owner;
	m_impl->requestId= requestId;
	m_impl->future= promise.get_future();
}

MikanResponseFuture::MikanResponseFuture(MikanResponseFuture&& other) noexcept
	: m_impl(new MikanResponseFutureImpl())
{
	m_impl->ownerRequestManager = other.m_impl->ownerRequestManager;
	m_impl->requestId = other.m_impl->requestId;
	m_impl->future = std::move(other.m_impl->future);
}

MikanResponseFuture::~MikanResponseFuture()
{
	if (m_impl != nullptr)
	{
		delete m_impl;
	}
}

bool MikanResponseFuture::isCompleted() const
{
	if (isValid())
	{
		// This is apparently the only way to check if a future is ready in a "non-blocking" way
		// https://stackoverflow.com/questions/10890242/get-the-status-of-a-stdfuture
		return m_impl->future.wait_for(std::chrono::seconds(0)) == std::future_status::ready;
	}
	else
	{
		return false;
	}
}


bool MikanResponseFuture::tryFetchResponse(MikanResponsePtr& outResponse)
{
	if (isCompleted())
	{
		outResponse = m_impl->future.get();
		return true;
	}

	return false;
}

MikanResponsePtr MikanResponseFuture::fetchResponse(uint32_t timeoutMilliseconds)
{
	if (isValid())
	{
		if (timeoutMilliseconds > 0)
		{
			// Wait for a fixed timeout
			if (m_impl->future.wait_for(std::chrono::milliseconds(timeoutMilliseconds)) == std::future_status::ready)
			{
				return m_impl->future.get();
			}
			else
			{
				// Timeout reached, cancel the request
				if (m_impl->ownerRequestManager != nullptr &&
					m_impl->requestId != INVALID_MIKAN_ID)
				{
					m_impl->ownerRequestManager->cancelRequest(m_impl->requestId);
				}

				// Return a timeout response instead
				return makeSimpleMikanResponse(MikanAPIResult::Timeout);
			}
		}
		else
		{
			return m_impl->future.get();
		}
	}
	else
	{
		return makeSimpleMikanResponse(MikanAPIResult::Uninitialized);
	}
}

void MikanResponseFuture::awaitResponse(uint32_t timeoutMilliseconds)
{
	// Drop the response on the floor
	fetchResponse(timeoutMilliseconds);
}

MikanResponsePtr MikanResponseFuture::makeSimpleMikanResponse(MikanAPIResult result)
{
	auto response = std::make_shared<MikanResponse>();
	response->requestId = INVALID_MIKAN_ID;
	response->resultCode = result;

	return response;
}

File: bindings/csharp/MikanResponseFuture.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace MikanXR
{
	public class MikanResponseFuture
	{		
		private MikanRequestManager _ownerRequestManager = null;
		private int _requestId = (int)MikanConstants.InvalidMikanID;
		private Task<MikanResponse> _future = null;

		public MikanResponseFuture()
		{
		}

		public MikanResponseFuture(MikanAPIResult result)
		{
			// Immediately set the result on the future
			_future = Task.FromResult(makeSimpleMikanResponse(result));
		}

		public MikanResponseFuture(
			MikanRequestManager owner,
			int requestId,
			TaskCompletionSource<MikanResponse> promise)
		{
			_ownerRequestManager= owner;
			_requestId= requestId;
			_future= promise.Task;
		}

		public MikanResponseFuture(MikanResponseFuture other)
		{
			_ownerRequestManager = other._ownerRequestManager;
			_requestId = other._requestId;
			_future = other._future;
		}

		public bool IsValid()
		{ 
			return _future != null; 
		}

		public bool IsCompleted()
		{
			return IsValid() && _future.IsCompleted;
		}

		// Non-Blocking Response Fetch
		// Return true if the response is ready, false otherwise
		public bool TryFetchResponse(out MikanResponse response)
		{
			response = null;

			if (IsCompleted())
			{
				response = _future.Result;
				return true;
			}

			return false;
		}

		// Blocking Response Fetch
		// Return the response if it is ready, or a timeout response if the timeout is reached
		public MikanResponse FetchResponse(double timeoutMilliseconds= 1000)
		{
			if (IsValid())
			{
				if (timeoutMilliseconds > 0)
				{
					// Wait for a fixed timeout
					_future.Wait(TimeSpan.FromMilliseconds(timeoutMilliseconds));

					if (_future.IsCompleted && !_future.IsFaulted)
					{
						return _future.Result;
					}
					else
					{
						// Timeout reached, cancel the request
						if (_ownerRequestManager != null &&
							_requestId != (int)MikanConstants.InvalidMikanID)
						{
							_ownerRequestManager.CancelRequest(_requestId);
						}

						// Return a timeout response instead
						return makeSimpleMikanResponse(MikanAPIResult.Timeout);
					}
				}
				else
				{
					// Wait indefinitely
					_future.Wait();

					// Return the result once the task succeeds or fails
					return _future.Result;
				}
			}
			else
			{
				return makeSimpleMikanResponse(MikanAPIResult.Uninitialized);
			}
		}

		// Blocking Response Await
		// Returns once the response has been received or the timeout is reached
		public void AwaitResponse(double timeoutMilliseconds= 1000)
		{
			// Drop the response on the floor
			FetchResponse(timeoutMilliseconds);
		}

		static MikanResponse makeSimpleMikanResponse(MikanAPIResult result)
		{
			var response = new MikanResponse();
			response.responseTypeId = MikanResponse.classId;
			response.responseTypeName = typeof(MikanResponse).Name;
			response.requestId = (int)MikanConstants.InvalidMikanID;
			response.resultCode = result;

			return response;
		}

	}
}
